File: src/morie/fn/aitamg.py
# morie.fn -- function file (rootcoder007/morie)
"""Amalgamation: sum a subset of parts into a single part."""

from . import _tail1core as C

from ._richresult import RichResult

__all__ = ["amalgam", "aitchison_amalgamation"]


def amalgam(x, parts, total=1.0):
    """Amalgamate a subset of parts into one, keeping the rest.

    Amalgamation is the other way to reduce the dimension of a
    composition, and unlike subcomposition it is NOT a log-ratio
    operation: it adds raw parts, so it does not commute with
    perturbation and it is not subcompositionally coherent.  It is
    included because real classification schemes amalgamate all the
    time, not because the geometry likes it.

    The amalgamated part is appended last; the retained parts keep
    their original order.  ``parts`` is ONE-BASED, matching the R arm.

    Formula: amalg(x; S) = C( (x_i : i not in S), sum_{j in S} x_j )

    Parameters
    ----------
    x : array-like
        Strictly positive vector of parts.
    parts : sequence of int
        One-based indices of the parts summed together.
    total : float
        Constant the result sums to.

    Returns
    -------
    RichResult
        ``composition``, ``amalgamated``, ``parts``, ``kept``, ``D``.

    References
    ----------
    Aitchison (1986), The Statistical Analysis of Compositional Data,
    Chapter 2, which defines amalgamation and notes that it does not
    preserve the log-ratio structure.
    """
    x = C.vec(x)
    if any(v <= 0 for v in x):
        raise ValueError("compositions must be strictly positive")
    D = len(x)
    idx = [int(i) for i in parts]
    if len(idx) < 2:
        raise ValueError("an amalgamation needs at least two parts")
    if any(not 1 <= i <= D for i in idx):
        raise ValueError("parts must be one-based indices in 1..D")
    sel = set(idx)
    keep = [i for i in range(1, D + 1) if i not in sel]
    amal = sum(x[i - 1] for i in idx)
    raw = [x[i - 1] for i in keep] + [amal]
    s = sum(raw)
    k = float(total)
    return RichResult(payload={
        "composition": [k * v / s for v in raw], "amalgamated": amal,
        "parts": idx, "kept": keep, "D": len(raw),
        "method": "Amalgamation"})


aitchison_amalgamation = amalgam


def cheatsheet():
    return "aitamg: amalg(x; S) = C(x_notS, sum x_S)"

File: r-package/morie/R/aitamg.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Amalgamation of a composition into grouped parts.
#'
#' Formula: amalg(x; g)_k = sum_{i : g_i = k} x_i, then closed
#'
#' @param x Composition with strictly positive parts.
#' @param groups Group label 1..k for each part of x, same length as x; every label from 1 to k must be used.
#' @param total Constant kappa the closure sums to.
#'
#' @return List with ``composition``, ``raw``, ``k``, ``total``, ``D``.
#' @references Aitchison, J. (1986), The Statistical Analysis of Compositional Data, Chapman and Hall, is this shelf's primary book and is NOT in the reference library, so it could not be read; see EXTERNAL_SOURCES.md.  Amalgamation adds parts together, unlike subcomposition which drops them.  It is the operation that is NOT subcompositionally coherent: log-ratios among the amalgamated groups are not determined by the log-ratios among the original parts, which is why the two operations are kept as separate functions here.  Group labels are 1-based in BOTH language arms.  Implemented in the standard published form.  The log-ratio algebra it rests on was verified against Mateu-Figueras, Pawlowsky-Glahn and Egozcue, arXiv:0802.2643 Sect. 4.1 (fetched and archived), but this particular definition is not printed there and could not be checked against Aitchison's own text.
#' @export
Compamalg <- function(x, groups, total = 1) {
  x <- .t1_vec(x); g <- as.integer(groups)
  if (length(g) != length(x)) stop("groups must have one label per part of x")
  if (length(x) == 0L) stop("x must be non-empty")
  if (any(x <= 0)) stop("compositions must be strictly positive")
  k <- max(g)
  if (min(g) != 1L || !identical(sort(unique(g)), seq_len(k)))
    stop("group labels must be 1..k with every label used")
  raw <- as.numeric(tapply(x, factor(g, levels = seq_len(k)), sum))
  s <- sum(raw); t <- as.numeric(total)
  .t1_result(composition = t * raw / s, raw = raw, k = k, total = t,
             D = length(x), method = "Amalgamation")
}
